compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -453,7 +453,17 @@ impl<'p, 'tcx> Matrix<'p, 'tcx> {
     /// expands it.
     fn push(&mut self, row: PatStack<'p, 'tcx>) {
         if !row.is_empty() && row.head().is_or_pat() {
-            self.patterns.extend(row.expand_or_pat());
+            let pats = row.expand_or_pat();
+            let mut no_inner_or = true;
+            for pat in pats {
+                if !pat.is_empty() && pat.head().is_or_pat() {
+                    self.patterns.extend(pat.expand_or_pat());
```

Comment:
Won't this just fall on its head again if you add a third level of or pattern?

Reply:
One of the tests should have covered the third level, but I just made a recursive function so it should be okay now.